hoomd/write/hdf5.py

```diff
@@ -0,0 +1,295 @@
+                raise RuntimeError(
+                    "The logged quantities cannot change within a file.")
+            dataset = self._fh[str_key]
+            dataset.resize(self._frame + 1, axis=0)
+            dataset[self._frame, ...] = value
+        self._frame += 1
+        self._fh["hoomd-data"].attrs["frames"] = self._frame
+
+    @_skip_fh
+    def flush(self):
+        """Write out all data currently buffered in memory.
+
+        Without calling this, data may be stored in the h5py.File object without
+        being written to disk yet.
+        """
+        self._fh.flush()
+
+    @_skip_fh
+    def _create_dataset(self, key: str, shape, dtype, chunk_size):
+        self._fh.create_dataset(
+            key,
+            shape,
+            dtype=dtype,
+            chunks=chunk_size,
+            maxshape=(None,) + shape[1:],
+        )
+
+    @_skip_fh
+    def _initialize_datasets(self, log_dict):
+        """Create datasets setting shape, dtype, and chunk size.
+
+        Chunk size does not seem to matter much to the write performance of the
+        file surprisingly. However, it has a drastic effect on reads. The
+        current settings balance file size (smaller scalar chunks) with reading
+        speed.
+
+        Tests were done on 1,000 frame files for writes and reads and tested for
+        writing and reading speed.
+        """
+        for key, (value, category) in log_dict.items():
+            chunk_size = None
+            if category == "scalar":
+                data_shape = (1,)
+                dtype = "f8" if isinstance(value, float) else "i8"
+                chunk_size = (self._SCALAR_CHUNK,)
+            else:
+                if not isinstance(value, np.ndarray):
+                    value = np.asarray(value)
+                data_shape = (1,) + value.shape
+                dtype = value.dtype
+                chunk_size = (max(
+                    self._MULTIFRAME_ARRAY_CHUNK_MAXIMUM // value.nbytes,
+                    1),) + data_shape[1:]
+            self._create_dataset("/".join(("hoomd-data",) + key), data_shape,
+                                 dtype, chunk_size)
+
+    @_skip_fh
+    def _find_frame(self):
+        if "hoomd-data" in self._fh:
+            return self._fh["hoomd-data"].attrs["frames"]
+        return 0
+
+    @_skip_fh
+    def _validate_scheme(self):
+        if "hoomd-data" in self._fh:
+            if "hoomd-schema" not in self._fh["hoomd-data"].attrs:
+                raise RuntimeError("Validation of existing HDF5 file failed.")
+        else:
+            group = self._fh.create_group("hoomd-data")
+            group.attrs["hoomd-schema"] = [0, 1]
+            group.attrs["frames"] = 0
+
+    def __getstate__(self):
+        state = copy.copy(self.__dict__)
+        del state["_fh"]
+        state["_attached_"] = False
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+
+
+class HDF5Logger(_InternalCustomWriter):
```

Comment:
`HDF5Logger` is awkward because it is _not_ a `Logger`. At the same time putting `HDF5` alongside `GSD` and `DCD` gives some hint that it might also write the system state when it does not.

What do you think of `HDF5Log` as a compromise solution?